### ISSUE TITLE
Bug: can't discard logs

### ISSUE THREAD

**Issue by d-nemi** [ozontech/file.d#681]
**Describe the bug**
Hi.
I've installed file.d as a daemonset in k8s cluster.
It works fine, but I can't filter logs that I don't need (like  `request_path: /api/status/liveness`).

Here's log example in pretty json (gathered from file.d pod log in debug mode):
```json
{
  "level": "info",
  "ts": "2024-10-01T14:22:39.065764723Z",
  "logger": "fd.k8s_kafka.action.debug",
  "message": "event sample",
  "offset": 4603887,
  "event": {
    "log": "{\"request_path\": \"/api/status/liveness\", \"response_body\": \"\\\"OK\\\"\", \"response_headers\": {\"content-length\": \"4\", \"content-type\": \"application/json\", \"x-request-id\": \"e1afa984-c91e-4ddc-80be-86f29a2471e8\"}, \"status_code\": 200, \"elapsed_ms\": \"0.85ms\", \"logger\": \"det_logger.request_logger\", \"request_id\": \"e1afa984-c91e-4ddc-80be-86f29a2471e8\", \"datetime\": \"2024-10-01T14:22:33.803672Z\", \"message\": \"Processed request with request_id=e1afa984-c91e-4ddc-80be-86f29a2471e8\", \"source_context\": \"/srv/www/.venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py:859\", \"log_level\": \"info\"}\n",
    "time": "2024-10-01T14:22:33.804016427Z",
    "stream": "stdout",
    "k8s_node": "01-x25",
    "k8s_namespace": "proxy-service",
    "k8s_pod": "proxy-service-8asdf4fd-v9gas",
    "k8s_container": "proxy-service",
    "k8s_pod_label_app.kubernetes.io/name": "proxy-service",
    "k8s_pod_label_pod-template-hash": "86574684fd",
    "k8s_pod_label_app.kubernetes.io/instance": "test",
    "k8s_node_label_kubernetes.io/arch": "amd64",
    "k8s_node_label_node.kubernetes.io/instance-type": "vsphere-vm.cpu-4.mem-16gb.os-linux",
    "k8s_node_label_node.kubernetes.io/role": "worker",
    "k8s_node_label_beta.kubernetes.io/arch": "amd64",
    "k8s_node_label_beta.kubernetes.io/instance-type": "vsphere-vm.cpu-4.mem-16gb.os-linux",
    "k8s_node_label_beta.kubernetes.io/os": "linux",
    "k8s_node_label_kubernetes.io/hostname": "01-x25",
    "k8s_node_label_kubernetes.io/os": "linux",
    "k8s_node_label_node-role.kubernetes.io/worker": "",
    "k8s_node_label_node.cluster.x-k8s.io/esxi-host": "s-asf4-t-esxd1.ex.com"
  }
}
```
I want to discard logs containing `request_path:/api/status/{liveness,readiness}`

Here's my configmap:

```yaml
  pipelines:
    k8s_kafka:
      input:
        type: k8s
        file_config:
          persistence_mode: async
          filename_pattern: "*"
        offsets_file: /data/k8s-offsets.yaml
      actions:
         - type: debug
            interval: 1s
            first: 10
        - type: rename
          override: true
          k8s_pod_label_app: "kubernetes.labels.pod_app"
        - type: json_decode
          field: log
          prefix: "json."
        - type: modify
          logger.name: "file-d"
          application_name: "My-app"
          team.name: "My-team"
          environment: "test"
          kubernetes_cluster: "my.test.cluster.com"
        - type: discard
          match_fields:
            k8s_namespace:
              - default
              - logging
            request_path: ^\/api\/status\/*
            match_mode: or

      output:
        type: kafka
        brokers: [my.broker-1.com:9092, my.broker-2.com:9092]
        default_topic: my-topic

```

I expect that logs with `request_path: /api/status/liveness` would be filtered, but they are still sent to kafka.
However logs with `k8s_namespace: default` are filtered as expected.

**What I've tried so far**
- changing regex, like 
`request_path: ^/api/status/*`
or more explicit:
```
request_path:
  - "/api/status/liveness"
  - "/api/status/readiness"
```
- adding a separate discard entry, e.g. before renaming:
```
          - type: discard
            match_fields:
              json\.request_path: 
                - "/api/status/liveness"
```
- adding do_if:
```
          - type: discard
            do_if:
              op: "or"
              operands:
              - op: "equal"
                field: "request_path"
                values: ["/api/status/liveness", "/api/status/readiness"]
```
- adding  one more key 'log':
```
          - type: discard
            match_fields:
              k8s_namespace:
                - default
                - logging
              log: 
                request_path: ^\/api\/status\/*
            match_mode: or
```

None of that helped.

**Troubleshooting**
I'd be happy to troubleshoot issue myself and found https://github.com/ozontech/file.d/blob/master/docs/configuring.md#actions-debugging
I'd like to add here, that `-http=':9090'` is invalid, correct option is `--http=:9000`
And after applying this configuration I get just:
```
curl http://1.2.3.4:9000/pipelines/k8s_kafka/2/sample
{"error":"There are no active processors"}

 curl -I http://1.2.3.4:9000/pipelines/k8s_kafka/2/sample
HTTP/1.1 400 Bad Request
```
Changing digit, like 1/sample, 3/sample won't help.

**Additional context**
Version: Hi! I'm file.d version=v0.34.0
Platform: k8s

--
I can provide additional info if needed.
Hope to get answer soon.

**Comment by HeadHunter483** [comment]
Hello! Thanks for submitting!

I've checked the configs with the provided log. In the original configmap there are several issues:

1. `match_mode: or` should be on the same level as `match_fields`, not inside of it
2. the compared value in `match_fields` is considered a regular expression only if it starts and ends by `/`, so the value for `request_path` should be `/^\/api\/status\/*/` to execute a regular expression
3. after applying `json_decode` plugin with `prefix: "json."`, your field `request_path` inside `log` transforms into `json. request_path`

Considering all of the above the condition for the `discard` should be:

```yaml
        - type: discard
          match_fields:
            k8s_namespace:
              - default
              - logging
            json\.request_path: /^\/api\/status\/*/
          match_mode: or
```

Note: since your field has a dot in its name the dot should be shielded by backslash `\` because usually it means a path to nested field. E.g. path `field.subfield` is treated as `{"field": {"subfield": "val"} }`.

Config 2

```yaml
request_path:
  - "/api/status/liveness"
  - "/api/status/readiness"
```

Did not work because of field name, it should be `json\.request_path`.

Config 3

```yaml
          - type: discard
            match_fields:
              json\.request_path: 
                - "/api/status/liveness"
```
           
Might've not worked because the logs you were testing it on did not meet the condition.

Config 4

```yaml
          - type: discard
            do_if:
              op: "or"
              operands:
              - op: "equal"
                field: "request_path"
                values: ["/api/status/liveness", "/api/status/readiness"]
```

Did not work because of field name.

In config 5

```yaml
          - type: discard
            match_fields:
              k8s_namespace:
                - default
                - logging
              log: 
                request_path: ^\/api\/status\/*
            match_mode: or
```

It doesn't work this way. If you need to point a path to field it should be `log.request_path`. In this particular case it wouldn't help though. The field `log` is removed after applying `json_decode` on it, and from the beginning its value was string, not object.

---

About troubleshooting with `http://1.2.3.4:9000/pipelines/k8s_kafka/2/sample`. The handler for this endpoint collects currently processed data in the action. I assume at the time of request there were no events in processing.

---

Also we recommend discard logs as early as possible for better performance. So the config could look like this:

```yaml
  pipelines:
    k8s_kafka:
      input:
        type: k8s
        file_config:
          persistence_mode: async
          filename_pattern: "*"
        offsets_file: /data/k8s-offsets.yaml
      actions:
        - type: debug
          interval: 1s
          first: 10
        - type: discard
          metric_name: discarded_k8s
          do_if:
            op: "equal"
            field: k8s_namespace
            values: ["default", "logging"]
        - type: json_decode
          field: log
          prefix: "json."
        - type: discard
          metric_name: discarded_parsed
          do_if:
            op: "equal"
            field: json\.request_path
            values: ["/api/status/liveness", "/api/status/readiness"]
        - type: rename
          override: true
          k8s_pod_label_app: "kubernetes.labels.pod_app"
        - type: modify
          logger.name: "file-d"
          application_name: "My-app"
          team.name: "My-team"
          environment: "test"
          kubernetes_cluster: "my.test.cluster.com"

      output:
        type: kafka
        brokers: [my.broker-1.com:9092, my.broker-2.com:9092]
        default_topic: my-topic
```

There are two `discard` plugins now, the first discard before decoding json from `log` to save some CPU and memory, the second discard by parsed field is right after `json_decode` to save resources on further actions. Also there are two metrics for `discard` actions, so you can track whether it discards logs or not. Both metrics are counters with label `status`. If you need additional labels, you can add them using field `metric_labels`.

**Comment by d-nemi** [comment]
Hi, @HeadHunter483!

First of all thanks for such a lightning-fast answer!
Feels like I've bought a paid subscription, not installed an open-source tool :)

Regarding your answer:
TL;DR it works! 
```
        - type: discard
          metric_name: discarded_parsed
          do_if:
            op: "equal"
            field: json\.request_path
            values: ["/api/status/liveness", "/api/status/readiness"]
```
this block helped.

---- 
Thanks for the explanations.
I've searched through documentation but couldn't find explicit info about slashes in regex, escaping dots in keys, etc.
I think it's a good idea to add it to docs, and I can bring a PR. Please suggest an appropriate file/place for that. 
Maybe just some more [examples](https://github.com/ozontech/file.d/blob/master/docs/examples.md)?


------
> About troubleshooting with http://1.2.3.4:9000/pipelines/k8s_kafka/2/sample. The handler for this endpoint collects currently processed data in the action. I assume at the time of request there were no events in processing.

Ok, I see.
What's the correct way to troubleshoot via endpoint?
I've tried `while true; do curl 1.2.3.4:9000/...; done` - but it didn't help, though logs were processed while querying the endpoint.
Maybe file.d is too fast and at the exact moment I do curl nothing is processed?

**Comment by HeadHunter483** [comment]
@d-nemi yeah I think examples is a good place. We have plans on reworking `match_fields` a bit to make it work together with `do_if`, so we'll add some more info about dots there too.

About troubleshooting, if you are testing locally by writing one or two logs manually, it could be useful to set more `debug` plugins right before and after the plugin you are interested in. Also metrics can help, they can be added to any action plugin using `metric_name`, and you can add values from the log as metric labels by providing field names in array `metric_labels`.

In that case with `match_fields` and `do_if`, you could test by replacing `discard` plugin with `modify` plugin. E.g.

```
        - type: modify
          shouldDiscard: "true"
          do_if:
            op: "equal"
            field: json\.request_path
            values: ["/api/status/liveness", "/api/status/readiness"]
```

**Comment by d-nemi** [comment]
@HeadHunter483, 
Thanks for your help!

I'll close the issue.
